server/datastore/mysql/migrations/tables/20230721161508_QueriesDataMigrator.go

```diff
@@ -0,0 +1,458 @@
+	type QueryWithTeamIDs struct {
+		fleet.Query
+		TeamRoles string
+	}
+
+	selectStmt := `
+		SELECT DISTINCT q.name,
+						q.description,
+						q.query,
+						q.author_id,
+						q.saved,
+						q.observer_can_run,
+						GROUP_CONCAT(CONCAT(ut.team_id, ':', ut.role)) AS team_roles
+		FROM queries q
+				LEFT JOIN scheduled_queries sq ON q.team_id IS NULL AND q.name = sq.query_name
+				INNER JOIN user_teams ut on q.author_id = ut.user_id
+		WHERE sq.id IS NULL
+		GROUP BY q.id`
+
+	rows, err := tx.Query(selectStmt)
+	if err != nil {
+		return fmt.Errorf("error executing 'Query' for non-scheduled queries: %s", err)
+	}
+	defer rows.Close()
+
+	var args []interface{}
+	var nRows int
+
+	for rows.Next() {
+		query := QueryWithTeamIDs{}
+		if err := rows.Scan(
+			&query.Name,
+			&query.Description,
+			&query.Query.Query,
+			&query.AuthorID,
+			&query.Saved,
+			&query.ObserverCanRun,
+			&query.TeamRoles,
+		); err != nil {
+			return fmt.Errorf("error executing 'Scan' for non-scheduled queries: %s", err)
+		}
+		teamRoles := strings.Split(query.TeamRoles, ",")
+		for _, teamRole := range teamRoles {
+			teamRoleParts := strings.Split(teamRole, ":")
+
+			role := teamRoleParts[1]
+			if role == "observer" || role == "observer_plus" {
+				continue
+			}
+			nRows += 1
+			teamID, err := strconv.Atoi(teamRoleParts[0])
+			if err != nil {
+				return fmt.Errorf("error parsing team ID on non-scheduled queries: %s", err)
+			}
+			args = append(args,
+				query.Name,
+				query.Description,
+				query.Query.Query,
+				query.AuthorID,
+				query.Saved,
+				query.ObserverCanRun,
+				teamID,
+				teamRoleParts[0],
+			)
+		}
+	}
+	if err := rows.Err(); err != nil {
+		return fmt.Errorf("error on 'rows' for non-scheduled queries: %s", err)
+	}
+	if err := rows.Close(); err != nil {
+		return fmt.Errorf("error executing 'Close' for non-scheduled queries: %s", err)
+	}
+
+	if len(args) == 0 {
+		return nil
+	}
+
+	insertStmt := `
+		INSERT INTO queries (
+			name,
+			description,
+			query,
+			author_id,
+			saved,
+			observer_can_run,
+			team_id,
+			team_id_char
+		) VALUES %s`
+
+	placeHolders := strings.TrimSuffix(strings.Repeat("( ?, ?, ?, ?, ?, ?, ?, ? ),", nRows), ",")
+	if _, err = tx.Exec(fmt.Sprintf(insertStmt, placeHolders), args...); err != nil {
+		return fmt.Errorf("error executing 'Exec' on non-scheduled queries: %s", err)
+	}
+	return nil
+}
+
+func _20230719152138_clean_up(tx *sql.Tx) error {
+	// Remove query stats
+	if _, err := tx.Exec(`TRUNCATE scheduled_query_stats`); err != nil {
+		return fmt.Errorf("error truncating 'scheduled_query_stats': %s", err)
+	}
+	if _, err := tx.Exec(`DELETE FROM aggregated_stats WHERE type = 'query' OR type = 'scheduled_query'`); err != nil {
+		return fmt.Errorf("error removing aggregated_stats: %s", err)
+	}
+
+	// Delete queries that only belong to 'global' and 'team' packs
+	if _, err := tx.Exec(`
+DELETE
+FROM queries
+WHERE name IN (SELECT query_name
```

Comment:
So if I understand correctly, the idea here is that only the old (pre-migration) queries would potentially get deleted, because the newly created ones (via the migration) have the "name - id" format and are not referenced by any existing pack, right?

What happens to packs and scheduled queries rows after the migration, since we don't remove them? I haven't seen the API/backend code changes yet, but a big picture-type summary, we only support "2017 packs" and those still have the old style schedules and queries? Global or team packs are fully migrated to the new queries and the corresponding `packs` and `scheduled_querues` entries are unused from then on?

Reply:
> So if I understand correctly, the idea here is that only the old (pre-migration) queries would potentially get deleted, because the newly created ones (via the migration) have the "name - id" format and are not referenced by any existing pack, right?

Yes, and because the newly created queries have all the scheduling information, while old ones don't.

> What happens to packs and scheduled queries rows after the migration, since we don't remove them? I haven't seen the API/backend code changes yet, but a big picture-type summary, we only support "2017 packs" and those still have the old style schedules and queries? Global or team packs are fully migrated to the new queries and the corresponding packs and scheduled_querues entries are unused from then on?

Yeah, the goal is to complete deprecate `global` and `team packs`. We actually delete all Global and Team packs on the next line, because the `pack_id` FK on the `scheduled_queries` has an `ON CASCADE DELETE` clause, removing those packs will also remove the related `scheduled_queries` rows.

Reply:
Oh, thanks, somehow I had completely missed the following code block that deletes packs :D ! Good to know that this also deletes scheduled queries.